Add EIP: Uncouple execution payload from beacon block (#9432)
ethereum/EIPs: EIPS/eip-uncouple-payload.md

---
title: Uncouple execution payload from beacon block
description: Separates the execution payload from beacon block to independently transmit them
author: Gajinder Singh <@g11tech>
discussions-to: https://ethereum-magicians.org/t/uncouple-execution-payload-from-beacon-block/23029
status: Draft
type: Standards Track
category: Core
created: 2025-03-01
---

## Abstract

> Currently the beacon block in ethereum consensus embeds the transactions in `ExecutionPayload` field of `BeaconBlockBody`. This EIP proposes to replace `ExecutionPayload` with `ExecutionPayloadHeader` in `BeaconBlockBody` and to independently transmit `ExecutionPayloadWithInclusionProof`.

Reviewer: ```suggestion
Currently, the beacon block in Ethereum Consensus embed transactions within `ExecutionPayload` field of `BeaconBlockBody`. This EIP proposes to replace `ExecutionPayload` with `ExecutionPayloadHeader` in `BeaconBlockBody` and to independently transmit `ExecutionPayloadWithInclusionProof`.
```

> However this EIP makes no change to the block import mechanism except that block availability also now includes waiting for the availability of `ExecutionPayloadWithInclusionProof` and this is where its different (and simpler) from proposals like ePBS.

Reviewer: ```suggestion
However, this EIP makes no change to the block import mechanism, with the exception that block availability now includes waiting for the availability of `ExecutionPayloadWithInclusionProof`, making it different and simpler from proposals like ePBS.
```

> However this EIP makes no change to the block import mechanism except that block availability also now includes waiting for the availability of `ExecutionPayloadWithInclusionProof` and this is where its different (and simpler) from proposals like ePBS.

Reviewer 2: ```suggestion
However this EIP makes no change to the block import mechanism except that block availability also now includes waiting for the availability of `ExecutionPayloadWithInclusionProof` and this is where it is different (and simpler) from proposals like ePBS.
```

> But this availability requirement can infact be restricted to `gossip` import while allowing optimistic syncing of ELs on checkpoint/range sync as EL can pull full blocks from their peers in optimistic sync as they do now.

Reviewer: ```suggestion
But this availability requirement can infact be restricted to `gossip` import while allowing optimistic syncing of the execution layer (EL) on checkpoint/range sync as EL can pull full blocks from their peers in optimistic sync as they do now.
```

## Motivation

> Ethereum protocol has an ambitious goal to grow the `gasLimit` of the execution payloads (possibly by 10X). This introduces bigger and bigger messages in the networking and  block processing pipeline of the CL clients leading to following issues:

Reviewer: ```suggestion
Ethereum protocol has an ambitious goal to grow the `gasLimit` of the execution payloads (possibly by 10X). This leads to larger messages, negatively affecting the networking and block processing pipelines of the consensus layer (CL) clients leading to following issues:
```

> 1. Latency for the arrival of Beacon Block increases and make it more and more dependent on the bandwidth resources available to the running node.

Reviewer: ```suggestion
1. Higher latencies for the arrival of beacon blocks increase, requiring larger bandwidth resources to be made available for the beacon node.
```

> 2. The greater number/size of `transactions` directly increase the merkelization compute time directly affecting the import time of the block.

Reviewer: ```suggestion
2. The greater number and size of transactions directly increase the merkelization compute time, increasing the import time of the block.
```

> We know from timing games that the block import latency greatly affects a client's performance to make correct head attestations. With this EIP, block transmission and import pipeline will be unchocked and allows for greater flexibility in reception of huge `ExecutionPayloadWithInclusionProof` while the beacon block can undergo processing. 

Reviewer: ```suggestion
We know from timing games that the block import latency greatly affects a client's performance to make correct head attestations. With this EIP, block transmission and block import processes will be decongested, allowing for greater flexibility in receiving a larger `ExecutionPayloadWithInclusionProof`, while the beacon block can simultaneously undergo processing. 
```

> Infact EL clients can also independently participate in the propagation/pull of the execution blocks. That mechanism however can be independently developed and is not the part of this EIP.

Reviewer: ```suggestion
In addition, EL clients can also independently participate in forwarding and receiving larger execution blocks. That mechanism however can be independently developed and is out of scope for this EIP.
```

> Other side benefits of this EIP are that 

Reviewer: Additional benefits obtained from this EIP:

> Other side benefits of this EIP are that 

Reviewer 2: I think the minor benefit is that there is no need to maintain `BlindedBeaconBlock` and `BlindedBeaconBlockBody` container since everything will be blinded after this EIP so there will only be one type of beacon block

Author: thanks, added to the benefits :)

> - consensus clients don't need to store and serve blocks with transactions and provides greater efficiency and reduced resource requirement for running the nodes.

Reviewer: ```suggestion
- Consensus clients don't need to store and serve blocks with transactions, providing greater efficiency and reduced resource requirements for running a beacon node.
```

> - the PBS pipleline becomes more efficient by proposer transmitting the signed block directly to the p2p network while also submitting to the builder/relay for independent reveal of the `ExecutionPayloadWithInclusionProof`.

Reviewer: ```suggestion
- The proposer-builder separation (PBS) pipeline becomes more efficient by the proposer transmitting the signed block directly to the p2p network, while submitting to the builder/relay for the independent reveal of the `ExecutionPayloadWithInclusionProof`.
```

> - in future with ZK proof of the EL block execution, one could treat the transactions just like blobs leveraging DAS mechanisms for the availability without worrying about their execution validity.

Reviewer: ```suggestion
- In the future with zero-knowledge (ZK) proof of the EL block execution, nodes could treat the transactions similarly to blobs which leverage data availability sampling (DAS) mechanisms for available data without requiring confirmation about their execution validity.
```

Author: updated the language for more clarity, do checkout if its cleaner now :+1:

## Specification

- `ExecutionPayload` in the `BeaconBlockBody` is replaced by `ExecutionPayloadHeader`

> - `ExecutionPayloadWithInclusionProof` is computed by the proposer/builder and transmitted independently on a gossip channel.

Reviewer 2: Do we need new builder api for this?

Author: updated thanks :)

> - `ExecutionPayloadWithInclusionProof` is computed by the proposer/builder and transmitted independently on a gossip channel.

Reviewer: ```suggestion
- `ExecutionPayloadWithInclusionProof` is computed by the block proposer/builder and transmitted independently on a separate gossip channel.
```

Author: updated the language for more clarity, do checkout if its cleaner now 👍

> - data availability checks now also wait for availability of a `VALID` `ExecutionPayloadWithInclusionProof`

Reviewer: ```suggestion
- Data availability checks now MUST wait for availability of a `VALID` `ExecutionPayloadWithInclusionProof`
```

Author: updated the language for more clarity, do checkout if its cleaner now 👍

> - `newPayload` on engine api is modified to now accept `newPayloadHeader` and a new method for `newExecutionPayload` is introduced to check the vailidity of the `ExecutionPayload` recieved within `ExecutionPayloadWithInclusionProof`

Reviewer 2: This sentence is a little confusing.

So `newPayload` method is renamed to `newPayloadHeader` which verifies `ExecutionPayloadHeader` in the incoming blinded block, and a new method named `newExecutionPayload` which verifies the full payload from the sidecar? 

Does CL need to get both verified in order to complete a state transition?

Author: have tried making it simpler, do let know if its clearer now :+1:

> - `newPayload` on engine api is modified to now accept `newPayloadHeader` and a new method for `newExecutionPayload` is introduced to check the vailidity of the `ExecutionPayload` recieved within `ExecutionPayloadWithInclusionProof`

Reviewer: ```suggestion
- `newPayload` on the Engine API is modified to now accept `newPayloadHeader` and a new method for `newExecutionPayload` is introduced to check the validity of the `ExecutionPayload` received within `ExecutionPayloadWithInclusionProof`
```

Author: updated the language for more clarity, do checkout if its cleaner now 👍

> ELs can optionally introduce `getExecutionPayload` method (like `getBlobs`) to help fast recovery of execution payload from the EL p2p network but as noted above that mechanism could be independently speced and is not part of this EIP.

Author: updated the language for more clarity, do checkout if its cleaner now 👍
